Add notes and examples about non-intuitive `PathBuf::set_extension` behavior

tbu- opened the pull request:
Basically, passing the empty string will actually remove the extension instead of setting it to the empty string. This might change what is considered to be an extension. Additionally, passing an extension that contains dots will make the path only consider the last part of it to be the new extension.

compiler-errors commented on library/std/src/path.rs:
```diff
@@ -1411,17 +1411,28 @@ impl PathBuf {
+    /// Note that there are some exceptions when the parameter contains dots
+    /// (then the new [`self.extension`] will only be the last part), or if the
+    /// parameter is the empty string and the file stem also contains
+    /// intternal dots.
```
typo

tbu- replied:
Fixed.

compiler-errors commented on library/std/src/path.rs:
```diff
@@ -1411,17 +1411,28 @@ impl PathBuf {
     ///
+    /// # Caveats
+    ///
+    /// Note that there are some exceptions when the parameter contains dots
```
This caveats section is a bit vague, what exceptions are there?

tbu- replied:
The exceptions are enumerated below that line.

m-ou-se commented on library/std/src/path.rs:
```diff
@@ -1411,17 +1411,28 @@ impl PathBuf {
     ///
+    /// # Caveats
+    ///
+    /// Note that there are some exceptions when the parameter contains dots
+    /// (then the new [`self.extension`] will only be the last part), or if the
+    /// parameter is the empty string and the file stem also contains internal
+    /// dots.
```
This sentence is a bit vague. How about something like this?

```suggestion
    /// The new `extension` may contain dots and will be used in its entirety,
    /// but only the part after the final dot will be reflected in [`self.extension()`][Self::extension].
    /// See the examples below.
```

tbu- replied:
Thanks, added. Also added the part about empty parameter and file stem with internal dots which was missing from your wording.

tbu- committed:
Add notes and examples about non-intuitive `PathBuf::set_extension` behavior

Basically, passing the empty string will actually remove the extension
instead of setting it to the empty string. This might change what is
considered to be an extension. Additionally, passing an extension that
contains dots will make the path only consider the last part of it to be
the new extension.

tbu- commented:
@rustbot label +S-waiting-on-review -S-waiting-on-author

m-ou-se commented:
@bors r+

bors commented:
:pushpin: Commit 40916ef88f7809c5e8956a3917e00681ad25c30b has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se 40916ef88f7809c5e8956a3917e00681ad25c30b -->
<!-- homu: {"type":"Approved","sha":"40916ef88f7809c5e8956a3917e00681ad25c30b","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
